src/pip/_internal/operations/install/wheel.py

```diff
@@ -377,13 +378,128 @@ def save(self) -> None:
 
         zipinfo = self._getinfo()
 
-        with self._zip_file.open(zipinfo) as f:
-            with open(self.dest_path, "wb") as dest:
-                shutil.copyfileobj(f, dest)
+        global total
+
+        ### UNCOMMENT THIS BLOCK FOR MAIN BRANCH CODE AND COMMENT THE OTHER BLOCK
+        #start = time.perf_counter()
+        #with self._zip_file.open(zipinfo) as f:
+        #    if zipinfo.file_size > 0:
+        #        with open(self.dest_path, "wb") as dest:
+        #            shutil.copyfileobj(f, dest, min(zipinfo.file_size, 1048576))
+        #end = time.perf_counter()
+        #global total_master
+
+        ### UNCOMMENT THIS BLOCK FOR FIX BRANCH CODE AND COMMENT THE OTHER BLOCK
+        start = time.perf_counter()
+        with open(self.dest_path, "wb") as dest:
+            if zipinfo.file_size > 0:
+                extract(self._zip_file, zipinfo, dest)
+        end = time.perf_counter()
+
+        total += (end-start)
+        if zipinfo.file_size > 10000000 or zipinfo.filename.endswith("/RECORD"):
+            print("extracting {} took {:.9f} cumulative {:.9f}".format(zipinfo.filename, end-start, total))
 
         if zip_item_is_executable(zipinfo):
             set_extracted_file_to_default_mode_plus_executable(self.dest_path)
 
+total = 0.0
+total_header = 0.0
+import struct
+import zlib
+import time
+from zipfile import BadZipFile, crc32, ZIP_STORED, ZIP_DEFLATED
+from zipfile import _get_decompressor
+from zipfile import _MASK_COMPRESSED_PATCH, _MASK_STRONG_ENCRYPTION, _MASK_ENCRYPTED
+
+structFileHeader = "<4s2B4HL2L2H"
+stringFileHeader = b"PK\003\004"
+sizeFileHeader = struct.calcsize(structFileHeader)
+
+_FH_SIGNATURE = 0
+_FH_EXTRACT_VERSION = 1
+_FH_EXTRACT_SYSTEM = 2
+_FH_GENERAL_PURPOSE_FLAG_BITS = 3
+_FH_COMPRESSION_METHOD = 4
+_FH_LAST_MOD_TIME = 5
+_FH_LAST_MOD_DATE = 6
+_FH_CRC = 7
+_FH_COMPRESSED_SIZE = 8
+_FH_UNCOMPRESSED_SIZE = 9
+_FH_FILENAME_LENGTH = 10
+_FH_EXTRA_FIELD_LENGTH = 11
+
+def extract(zipfile, zipinfo, dest_fp):
+        zef_file = zipfile.fp
+        zef_file.seek(zipinfo.header_offset)
+
+        def check_header():
```

Comment:
code from the stdlib to check the header. (i've split to see the performance impact of that function in particular).
notice the few extra read() I/O calls to read the headers.